clients/da-vinci-client/src/main/java/com/linkedin/davinci/consumer/VeniceAfterImageConsumerImpl.java

```diff
@@ -166,47 +181,58 @@ protected CompletableFuture<Void> internalSeek(
     return super.internalSeek(partitions, targetTopic, seekAction);
   }
 
-  private class VersionSwapDetectionThread implements Runnable {
+  private class VersionSwapDataChangeListener implements StoreDataChangedListener {
     @Override
-    public void run() {
-      Set<Integer> partitions = new HashSet<>();
-      try {
-        // Check the current version of the server
-        int currentVersion = storeRepository.getStore(storeName).getCurrentVersion();
-
-        // Check the current ingested version
-        Set<PubSubTopicPartition> subscriptions = getTopicAssignment();
-        synchronized (subscriptions) {
-          if (subscriptions.isEmpty()) {
-            return;
-          }
-          int maxVersion = -1;
-          for (PubSubTopicPartition topicPartition: subscriptions) {
-            int version = Version.parseVersionFromVersionTopicName(topicPartition.getPubSubTopic().getName());
-            if (version >= maxVersion) {
-              maxVersion = version;
+    public void handleStoreChanged(Store store) {
+      synchronized (this) {
+        Set<Integer> partitions = new HashSet<>();
+        try {
+          // Check the current version of the server
+          int currentVersion = storeRepository.getStore(storeName).getCurrentVersion();
+
+          // Check the current ingested version
+          Set<PubSubTopicPartition> subscriptions = getTopicAssignment();
+          synchronized (subscriptions) {
+            if (subscriptions.isEmpty()) {
+              return;
+            }
+            int maxVersion = -1;
+            for (PubSubTopicPartition topicPartition: subscriptions) {
+              int version = Version.parseVersionFromVersionTopicName(topicPartition.getPubSubTopic().getName());
+              if (version >= maxVersion) {
+                maxVersion = version;
+              }
             }
-          }
 
-          // Seek to end of push
-          if (currentVersion != maxVersion) {
-            // get current subscriptions and seek to endOfPush
-            for (PubSubTopicPartition partitionSubscription: subscriptions) {
-              partitions.add(partitionSubscription.getPartitionNumber());
+            // Seek to end of push
+            if (currentVersion != maxVersion) {
```

Comment:
Will this work when we rollback current version?

Reply:
lol nope.

Actually on reading this code it's kind of whack.  It basically says that so long as there is any partition that is consuming on the correction version, leave the rest alone, otherwise, migrate all partitions.  So a lot of this should change. (all of this code was copy pasted from the original version swap code)